### ISSUE TITLE
Cross building

### ISSUE THREAD

**Issue by ittaiz** [bazelbuild/rules_scala#80]
So this isn't anything urgent but popped into my head and thought it would be best to start the conversation now before 2.12 is released and we might have more pressure.
2 questions:
1. How to parameterize the rule with respect to scala version? (don't think it's complicated but haven't looked into it).
2. How to cross build a target to both 2.11 and 2.12 like sbt supports. This is useful for both rolling out the upgrade in an organization and for OSS libraries.

**Comment by johnynek** [comment]
I think this is similar to #14 

I agree we should work on it.

The ideal case is that we just point at different repositories and maybe some functions to get the particular version numbers out.

We can certainly factor common code to something like `scala_common.bzl` and then call that from `scala2_11.bzl` and `scala2_12.bzl` etc...

I don't think it is *too* hard, I just have not taken a stab.

Probably `scala2_10.bzl` might be the first to do since many still use scala 2.10 and there are many libraries published that we could check with.

**Comment by johnynek** [comment]
this could be a useful approach: https://github.com/bazelbuild/rules_scala/pull/84

**Comment by ittaiz** [comment]
Indeed. I think one of the other rules uses it (rust? Go?)

**Comment by pauldraper** [comment]
https://github.com/bazelbuild/rules_scala/pull/84 doesn't let you cross build, right? It just gives you a way to globally change the version?

A couple more thoughts
1. Cross-building (as opposed to just choosing one version) is very important.
  a. Most popular Scala libraries cross-build against two or more versions of Scala.
  b. Even if a project is not for pubic consumption, it can take some time to migrate a large codebase and dependencies to a new Scala version. I used SBT + https://github.com/lucidsoftware/sbt-cross to do this during 2.10 -> 2.11, and it was very helpful.
2. The cross-built versions project can differ in more than just Scala version, e.g. some 2.11 projects need scala-xml, while in 2.10, this dependency was not separate.
3. It would be nice to keep just the one worker process, and use classloaders to load the different Scala compilers.

**Comment by johnynek** [comment]
I agree cross-building is somewhat interesting, but it is a very low priority for me personally. I'm happy to review PRs.

Bazel is generally used for private (logical) monorepos. There is no support for publishing, for instance. We could have multiple parallel builds, or we could say that the cross building solution is to have your CI bump the scala version you are using.

At Twitter, where we used the similar pants build system, there was no support for this, and in general keeping a large monorepo green with more than one version of scala, I think will be very challenging, not to mention tooling to support versioning the external jars correctly, something bazel does not understand (that scala has multiple parallel artifacts for each different release of scala).

**Comment by oscar-stripe** [comment]
see #251

**Comment by KaoruDev** [comment]
Checking in if this is something that will be supported. Otherwise, in order for us to adopt Bazel, we'll have to get _all_ of our services on the same Scala version (which isn't ideal).

**Comment by johnynek** [comment]
I think this thread is still the state of the art. You can switch scala versions and so you can manage to have your CI set two different versions, but having a two versions in the same build is not yet supported.

It sounds like you want to build some parts of the code with only version A and others with one version B. Is that right?

A key challenge is this: we want to support interoperability with java (java can depend on scala). But we can’t add two versions of the scala library on the class path. So we need some way to error if that happens.

I don’t think we have found a design where multiple versions of scala in the same build in bazel works great.

**Comment by KaoruDev** [comment]
SGTM, thanks for the tip on the CI switch

**Comment by harpocrates** [comment]
I wonder what Databricks did internally to support [`cross_scala_library` from this post](https://www.databricks.com/blog/2019/02/27/speedy-scala-builds-with-bazel-at-databricks.html). That looks exactly like what I was looking for, but based on this thread it doesn't look like this is supported here and I couldn't find any of their code in the open.

**Comment by eed3si9n** [comment]
Wrote a blog post [cross build anything with Bazel](https://eed3si9n.com/cross-build-anything-with-bazel/) documenting the stateful cross building (aka switch at CI approach).
